### PR TITLE
use Fire CLI

### PR DESCRIPTION
**Pull request by Borda**
simplification with Fire CLI and reduce duplicate code

### CODE REVIEW

File: run_on_azureml.py [hunk 1]
```diff
@@ -35,36 +34,31 @@ def find_or_create_compute_target(
...
+    exp = Experiment(ws, experiment_name)
+    target = find_or_create_compute_target(ws, compute_target_name)
+    env = Environment.from_pip_requirements("my-pip-env", requirements_file)
     run_config = ScriptRunConfig(
-        source_directory=Path(args.script_path).parent,
-        script=Path(args.script_path).name,
-        arguments=args.script_args,
+        source_directory=Path(script_path).parent,
+        script=Path(script_path).name,
+        arguments=kwargs,
```
Comment (nateraw) [inline run_on_azureml.py]: What are `kwargs` here? I assume its a dict. Does `ScriptRunConfig` accept dictionaries? I always thought it was supposed to be a list of strings (essentially whatever you want `sys.argv` to be in your script context).

File: run_on_azureml.py [hunk 2]
```diff
@@ -35,36 +34,31 @@ def find_or_create_compute_target(
...
-    target = find_or_create_compute_target(ws, args.compute_target_name)
-    env = Environment.from_pip_requirements("my-pip-env", args.requirements_file)
+def main(
+        workspace_config: str = "config.json",
+        experiment_name: str = "azureml-demo-exp",
+        compute_target_name: str = "my-compute",
+        requirements_file: str = "requirements.txt",
+        script_path: str = "simple_example/run.py",
+        delete_compute: bool = False,
+        **kwargs,
```
Comment (nateraw) [inline run_on_azureml.py:45]: Related to comment below -`extras` from `parser.parse_known_args(args)` is a list of strings, not a dict, so not same as kwargs.

File: run_on_azureml.py [hunk 3]
```diff
@@ -1,24 +1,23 @@
...
 
 def find_or_create_compute_target(
-    workspace,
-    name,
-    instance_type="Standard_D8_v3",
-    min_nodes=0,
-    max_nodes=10,
-    idle_seconds_before_scaledown=900,
-    vm_priority="lowpriority",
+        workspace,
```
Comment (nateraw) [inline run_on_azureml.py:15]: hehe why the double tab? is this something your formatter did?

File: simple_example/score.py [hunk 4]
```diff
@@ -38,4 +33,4 @@ def main(args):
 
 
 if __name__ == "__main__":
-    main(parse_args())
+    fire.Fire(main)
```
Comment (nateraw) [inline simple_example/score.py:36]: 🔥

**Review by nateraw — COMMENTED** [review]
Thank you so much for the contribution 😄 . Truly appreciate it. We gotta watch out for kwargs...not familiar with python fire, so I don't know how this works. 

- Will `**kwargs` accept any arbitrary flags passed in when executing from CLI when run with Python Fire?
- If so, does it work with argparse 'action'-type flags? (ex. `--do_train` without having to say `--do_train True`).


I already made a wrapper for this script in a separate repo, but I used `typer`, as I was able to use their context feature to get arbitrary additional script args. If Fire does this out of the box with **kwargs syntax, I'd much prefer that.

Here's what the typer syntax looks like...

```python
@app.command(
    name="run",
    context_settings={"allow_extra_args": True, "ignore_unknown_options": True},
)
def run(
    ctx: typer.Context,
    script: Path,
    workspace_config: Path = typer.Option(
        "./config.json",
        help="AzureML resource config.json file. Download from Azure portal by going to your resource's page and selecting 'Download config.json'",
    ),
    target_name: str = typer.Option(
        "my-cluster",
        help="The name of the compute cluster to run your script on. If it doesn't exist, we'll create it for you.",
    ),
    gpu_type: GPUType = typer.Option(
        GPUType.k80,
        help="The type of GPU you'd like to be on your cluster. Pass 'CPU' if you don't require a GPU.",
        case_sensitive=False,
    ),
    num_gpus: int = typer.Option(
        1,
        callback=num_gpus_callback,
        help="Number of GPUs per node. Can be 1, 2, or 4. If --gpu_type is CPU, this argument will be ignored, effectively setting it to 0.",
    ),
    min_nodes: int = 0,
    max_nodes: int = 10,
    num_nodes: int = 1,
    dataset_name: str = None,
    dataset_mount_dir: Path = "/dataset",
    experiment_name: str = "fuego-experiments",
    requirements_file: Path = "./requirements.txt",
    docker_image: str = "mcr.microsoft.com/azureml/openmpi3.1.2-cuda10.2-cudnn8-ubuntu18.04",
):

    assert script.exists(), f"Script file '{script}' does not exist"
    assert workspace_config.exists(), f"Workspace config file '{workspace_config}' does not exist"
    assert requirements_file.exists(), f"Requirements file '{requirements_file}' does not exist"

    submit_basic_run(
        script,
        ctx.args,
        workspace_config,
        target_name,
        gpu_type,
        num_gpus,
        min_nodes,
        max_nodes,
        num_nodes,
        dataset_name,
        str(dataset_mount_dir),
        experiment_name,
        requirements_file,
        docker_image,
    )
```

Here, `submit_basic_run` is essentially the same as our main function here, but is not shown. `ctx.args` is a list of strings representing extra args not parsed since they don't match variables defined in the signature of `run`.

You can see this syntax is a little annoying, as you'll still need two functions (since one will need to have 'ctx', which is annoying). If `fire` can do this, lets run with that instead 😄 .

**Comment by nateraw** [comment]
@Borda I added you as a contributor - keep in mind I have blogposts that reference this code. Don't let that stop you from making changes, but its perhaps worth mentioning in the readme that this code has diverged from the blogpost.

**Comment by Borda** [comment]
@nateraw thank you, I'll address your comments later this week
how about setting the only CI test as mandatory to prevent accident merging unfinished work... :rabbit:

**Comment by nateraw** [comment]
@Borda Not 100% sure what you mean by setting CI test as mandatory, but I just enabled a branch protection rule to enforce the tests pass before merging. Is that what you wanted? If not feel free to do it/explain what I need to do if you don't have access.

As for addressing the comments - **no rush**. This repo hasn't seen a real commit in months, so not concerned with how long it takes you to get to this 😄 . I have very limited bandwidth for this as well, so let's just take our time.

File: run_on_azureml.py (hunk 3 above)
Comment (Borda) [reply run_on_azureml.py:15]: seems so...

**Comment by Borda** [comment]
> We gotta watch out for kwargs...not familiar with python fire, so I don't know how this works.
> 
> * Will `**kwargs` accept any arbitrary flags passed in when executing from CLI when run with Python Fire?
> * If so, does it work with argparse 'action'-type flags? (ex. `--do_train` without having to say `--do_train True`).

just testing the **kwargs and seems to work as expected, see following example:
```py
import fire

def main(abc: int = 123, do: bool = False, **kwargs):
    print(f"abc: {abc}")
    print(f"do: {do}")
    print(f"kwargs: {kwargs}")

if __name__ == '__main__':
    fire.Fire(main)
```
and some call outputs:
```bash
jirka@jirka-Predator:~/Workingspace/azureml-examples$ python ssample.py 
abc: 123
do: False
kwargs: {}
jirka@jirka-Predator:~/Workingspace/azureml-examples$ python ssample.py --abc 456
abc: 456
do: False
kwargs: {}
jirka@jirka-Predator:~/Workingspace/azureml-examples$ python ssample.py --do --bc 5
abc: 123
do: True
kwargs: {'bc': 5}
```
also see: https://google.github.io/python-fire/guide/#functions-with-varargs-and-kwargs

File: run_on_azureml.py (hunk 1 above)
Comment (Borda) [reply run_on_azureml.py]: `kwargs` is dict so it shall be:
```py
    args = []
    for n, v in kwargs.items():
        args += [f"--{n}", f"{v}"]
```

File: deploy_on_azureml.py [hunk 5]
```diff
@@ -1,46 +1,41 @@
...
-    parser.add_argument("--model_artifact_path", type=str, default="logs/saved_model")
-    parser.add_argument("--score_file", type=str, default="keras_example/score.py")
-    parser.add_argument("--model_name", type=str, default="test-keras-model")
-    parser.add_argument("--service_name", type=str, default="test-keras-service")
-    return parser.parse_args(args)
-
-
-def main(args):
-
+def main(
```
Comment (nateraw) [inline deploy_on_azureml.py:11]: cute

File: run_on_azureml.py [hunk 6]
```diff
@@ -35,36 +35,32 @@ def find_or_create_compute_target(
...
+        requirements_file: str = "requirements.txt",
+        script_path: str = "simple_example/run.py",
+        delete_compute: bool = False,
+        **kwargs,
+):
+    ws = Workspace.from_config(workspace_config)
+    exp = Experiment(ws, experiment_name)
+    target = find_or_create_compute_target(ws, compute_target_name)
+    env = Environment.from_pip_requirements("my-pip-env", requirements_file)
+    args = list(itertools.chain(*[(f"--{n}", f"{v}") for n, v in kwargs.items()]))
```
Comment (nateraw) [inline run_on_azureml.py:51]: how does this handle bools?

Comment (nateraw) [reply run_on_azureml.py:51]: Non explicitly defined, I mean.

**Review by nateraw — APPROVED** [review]
LGTM thanks man 🍻

**Pull request merged**